### PR TITLE
Improves testing for `quest::STEPReader` for surface orientation

### PR DESCRIPTION
**Pull request by jcs15c**
# Summary

- This PR is a slight refactoring of the `quest::STEPReader` to handle additional edge cases stemming from different canonical representations of OpenCascade STEP files and `primal`'s trimmed NURBS primitives. In general, OCC stores surface orientation through both an internal flag and the orientation of loops of trimming curves, while `primal` encodes orientation information only through patch control points and requires CCW trimming curve loops.

An issue was identified and partially fixed in the closed PR #1765, but since then additional edge cases have popped up. To better test for these issues as the reader is exposed to more in-the-wild meshes, a dedicated test has been added to `quest`. This test is supported by additional STEP files added to [Axom's data repo](https://github.com/llnl/axom_data/pull/37).

Below is a table which shows which tested files fail for each configuration of the STEP reader. Files marked with * are new to `axom_data`.

| Filename | Pre #1765 | develop | current PR |
| -------- | ---------------- | ---------------- | ---------------- |
| tet        |            ✓              |          ✓          |          ✓         |
| sliced_cylinder |       ✓     |          ✓        |          ✓       |
| nut |           ✓      |       ✓        |          ✓         |
| bearings |      ✓     |        ✓        |     ✓      |
| boxed_sphere | |  ✓   |✓  | ✓     |
| plate* |    |   |✓ |
| bolt_clip* | ✓  | | ✓|

Additionally, this PR addresses an incorrect overload in `winding_number.hpp` for 2D NURBS curves.

### CODE REVIEW

File: src/axom/primal/operators/winding_number.hpp [hunk 1]
```diff
@@ -383,13 +383,19 @@ axom::Array<double> winding_number(const axom::Array<Point<T, 2>>& query_arr,
   {
     ret_val[n] = 0.0;
 
-    for(int i = 0; i < nurbs_curve_arr.size(); ++i)
+    for(int i = 0; i < nurbs_cache_arr.size(); ++i)
```
Comment (kennyweiss) [inline src/axom/primal/operators/winding_number.hpp]: Minor: should we be using the iterator interface on any/all/some of these loops to avoid worrying about indexing?

File: src/axom/quest/io/STEPReader.cpp [hunk 2]
```diff
@@ -66,6 +69,7 @@ struct PatchData
   int patchIndex {-1};
   bool wasOriginallyPeriodic_u {false};
   bool wasOriginallyPeriodic_v {false};
+  bool reversed_u {false};
```
Comment (kennyweiss) [inline src/axom/quest/io/STEPReader.cpp]: Suggestion on variable name (minor): 

Would `was_reversed_u` be clearer?

File: src/axom/quest/io/STEPReader.cpp [hunk 3]
```diff
@@ -867,14 +886,12 @@ class StepFileProcessor
         const TopoDS_Wire& wire = TopoDS::Wire(wireExp.Current());
 
         int edgeIndex = 0;
-        for(TopExp_Explorer edgeExp(wire, TopAbs_EDGE); edgeExp.More(); edgeExp.Next(), ++edgeIndex)
+        BRepTools_WireExplorer edgeExp(wire, TopoDS::Face(faceExp.Current()));
+        for(; edgeExp.More(); edgeExp.Next(), ++edgeIndex)
```
Comment (kennyweiss) [inline src/axom/quest/io/STEPReader.cpp:890]: Suggestion variable name (minor):

Since this is a `WireExplorer`, should the variable be `wireExp` instead of `edgeExp`?

File: src/axom/quest/tests/quest_step_reader.cpp [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+template <typename... Args>
+std::string pjoin(const char* str, Args... args)
+{
+  return fs::joinPath(std::string(str), pjoin(args...));
+}
+
+//------------------------------------------------------------------------------
+// Use a pared-down version of the TriangleGWN3D method to directly evaluate
+//  the for a triangle mesh
```
Comment (kennyweiss) [inline src/axom/quest/tests/quest_step_reader.cpp]: Missing word(s) in this comment. 

Perhaps:

```suggestion
//  the winding number for a triangle mesh
```

File: src/axom/quest/tests/quest_step_reader.cpp [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+    m_scale = shapeBBox.getMax()[longestDim] - shapeBBox.getMin()[longestDim];
+
+    const auto& ctr = m_shapeCenter;
+    const auto scl = m_scale;
+    auto trisView = m_triangles.view();
+    mint::for_all_cells<axom::SEQ_EXEC, mint::xargs::coords>(
+      &mesh,
+      AXOM_LAMBDA(axom::IndexType cellIdx,
+                  const axom::numerics::Matrix<double>& coords,
+                  [[maybe_unused]] const axom::IndexType* nodeIds) {
```
Comment (kennyweiss) [inline src/axom/quest/tests/quest_step_reader.cpp]: I think `[[maybe_unused]]` in a lambda was an error in some configs (e.g. hip) and the following fixed it:

```suggestion
                  const axom::IndexType* /* nodeIds */) {
```

File: src/axom/quest/tests/quest_step_reader.cpp [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+  EXPECT_GT(patches.size(), 0) << "No NURBS patches were extracted from '" << fileName << "'";
+  if(patches.empty())
+  {
+    return;
+  }
+
+  const auto shapeBbox = stepReader.getBRepBoundingBox().scale(1.1);
+  auto bboxMin = shapeBbox.getMin();
+  auto bboxMax = shapeBbox.getMax();
+  const auto bboxDiag = bboxMax.array() - bboxMin.array();
```
Comment (kennyweiss) [inline src/axom/quest/tests/quest_step_reader.cpp]: No need to change it, and the semantics are slightly different, but `primal::BoundingBox::range()` would have the same effect.

https://github.com/llnl/axom/blob/c84919a35b81631efdbe1dfdca1d4aef3a7accc3/src/axom/primal/geometry/BoundingBox.hpp#L156-L161

File: src/axom/quest/io/STEPReader.cpp (hunk 3 above)
Comment (jcs15c) [reply src/axom/quest/io/STEPReader.cpp:890]: Maybe, but `wireExp` is already defined above as an object of type `TopExp_Explorer` since that iterates over the topology, producing wires. Then the object of type `BRepTools_WireExplorer` iterates over the wires, producing edges. It's a little complicated, but I think there is logic to it. Wire explorers explore wires, and edge explorers explore edges.

File: src/axom/quest/tests/quest_step_reader.cpp (hunk 5 above)
Comment (rhornung67) [reply src/axom/quest/tests/quest_step_reader.cpp]: We also have this macro, which is preferred because it's easy to search for. https://github.com/llnl/axom/blob/develop/src/axom/core/Macros.hpp#L168
